engine/src/main/java/org/terasology/world/block/BlockLifecycleEvent.java

```diff
@@ -16,14 +16,18 @@
 package org.terasology.world.block;
 
 import gnu.trove.list.TIntList;
+import org.joml.Vector3i;
+import org.joml.Vector3ic;
 import org.terasology.entitySystem.event.Event;
-import org.terasology.math.geom.Vector3i;
 import org.terasology.world.BlockEntityRegistry;
-import org.terasology.world.block.internal.BlockPositionIterator;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
 
 /**
  */
-public abstract class BlockLifecycleEvent implements Event {
+public abstract class BlockLifecycleEvent implements Event, Iterable<Vector3ic> {
```

Comment:
```suggestion
//FIXME: There is a mismatch between `blockCount` and iterated blocks.
//       I don't get what this event is about, and how it should behave. The `TIntList positions` is a flattened list
//       of positions. So if this event affects
//          [(0,0,0), (1,2,3), (4,5,6)]
//       then `positions` would be:
//          [0,0,0, 1,2,3, 4,5,6]
//       But whether a block listed in positions is actually returned depends on whether there's a permanent block
//       entity registered or not? Doesn't this lead to a mismatch between the size as returned by blockCount and 
//       the actual amount of positions returned by the iterator? So, it could be that 
//          event.blockCount() == n 
//       but 
//          for (Vector3ic pos: event) { ... }
//       only iterates over m < n blocks? Maybe we should not inline the iterator here, but rather make it explicit
//       that it will only iterator over `blockPositionsWithNonPermanentEntities`.
public abstract class BlockLifecycleEvent implements Event, Iterable<Vector3ic> {
```